ipn/ipnlocal/node_backend.go

```diff
@@ -358,6 +362,36 @@ func (nb *nodeBackend) PeerAPIBase(p tailcfg.NodeView) string {
 	return peerAPIBase(nm, p)
 }
 
+// PeerIsReachable reports whether the current node can reach p. If the ctx is
+// done, this function may return a result based on stale reachability data.
+func (nb *nodeBackend) PeerIsReachable(ctx context.Context, p tailcfg.NodeView) bool {
+	if !nb.SelfHasCap(tailcfg.NodeAttrClientSideReachability) {
+		// Legacy behavior is to always trust the control plane, which
+		// isn’t always correct because the peer could be slow to check
+		// in so that control marks it as offline.
+		// See tailscale/corp#32686.
+		return p.Online().Get()
+	}
+
+	nb.mu.Lock()
+	nm := nb.netMap
+	nb.mu.Unlock()
+
+	if self := nm.SelfNode; self.Valid() && self.ID() == p.ID() {
+		// This node can always reach itself.
+		return true
+	}
+	return nb.peerIsReachable(ctx, p)
+}
+
+func (nb *nodeBackend) peerIsReachable(ctx context.Context, p tailcfg.NodeView) bool {
+	// TODO(sfllaw): The following does not actually test for client-side
+	// reachability. This would require a mechanism that tracks whether the
+	// current node can actually reach this peer, either because they are
+	// already communicating or because they can ping each other.
+	return p.Online().Get() || time.Since(p.LastSeen().Get()) <= GoneTooLong
```

Comment:
this is doing a subtraction different clicks: the client's clock (time.Since uses time.Now, the system clock) and the server's clock (p.LastSeen)

Reply:
The alternative that we had discussed earlier was a hammer: simply to `return true` if the flag were set.

If the clocks are reasonably close, then the client will be slightly more forgiving than the server. (I suppose this should also test if p.LastSeen is in the future.)

I also considered just keeping track, with a local clock, of the last map response where the peer was declared online. But then GoneTooLong will probably have to be on the order of minutes?

The next step is to wire in a call to [LocalBackend.Ping](https://pkg.go.dev/tailscale.com/ipn/ipnlocal#LocalBackend.Ping), but I think that change is a bit bigger and riskier since we need to:
1. Ping in parallel
2. Move the Ping implementation into nodeBackend.